### ISSUE TITLE
Index-SPI changes doesn't load "on-heap" dictionary as defined in FieldConfigList

### ISSUE THREAD

**Issue by vvivekiyer** [apache/pinot#12254]
### Issue
**Table Config**

```
{
  "tableName": "customer_OFFLINE",
  "tableType": "OFFLINE",
  "segmentsConfig": {
    "replication": "1",
    "segmentPushType": "REFRESH",
    "schemaName": "customer",
    "minimizeDataMovement": false
  },
  "tenants": {},
  "tableIndexConfig": {
    "autoGeneratedInvertedIndex": false,
    "enableDynamicStarTreeCreation": false,
    "columnMajorSegmentBuilderEnabled": false,
    "optimizeDictionaryForMetrics": false,
    "noDictionarySizeRatioThreshold": 0.85,
    "rangeIndexVersion": 2,
    "sortedColumn": [
      "C_NAME"
    ],
    "loadMode": "HEAP",
    "enableDefaultStarTree": false,
    "aggregateMetrics": false,
    "nullHandlingEnabled": false,
    "optimizeDictionary": false,
    "createInvertedIndexDuringSegmentGeneration": false
  },
  "metadata": {
    "customConfigs": {}
  },
  "fieldConfigList": [
    {
      "name": "C_NAME",
      "encodingType": "DICTIONARY",
      "indexTypes": [],
      "indexes": {
        "dictionary": {
          "disabled": false,
          "onHeap": true,
          "useVarLengthDictionary": true
        }
      },
      "tierOverwrites": null
    }
  ],
  "ingestionConfig": {
    "transformConfigs": [],
    "continueOnError": false,
    "rowTimeValueCheck": false,
    "segmentTimeValueCheck": true
  },
  "isDimTable": false
}
```

- Based on this config, we expect the dictionary for C_NAME column to be loaded on-heap. However, that doesn't happen. 

### RCA
- The issue is that `IndexLoadingConfig`  loads the configs for each column by calling  `createIndexConfigsByColName(tableConfig, schema, this::getDeserializer);` at [link](https://github.com/apache/pinot/blob/7132a2203f13478f450cbf8f0524ba72bdc267b7/pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/IndexLoadingConfig.java#L288) which uses the deserializer as defined [here](https://github.com/apache/pinot/blob/7132a2203f13478f450cbf8f0524ba72bdc267b7/pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/IndexLoadingConfig.java#L291).
- This deserializer provider (onConflict.PICK_FIRST). gives priority to  `DictionaryIndexType.fromIndexLoadingConfig()` over `fieldConfigs`. As `DictionaryIndexType.fromIndexLoadingConfig()` returns a DictionaryConfig with onHeap set to false. So the fieldConfig settings are never considered. 

### Questions/Solution
- As per my understanding of Index-SPI changes, `fieldConfigList` should be given higher priority over Index `tableIndexConfig`. Given this, shouldn't the code [here](https://github.com/apache/pinot/blob/7132a2203f13478f450cbf8f0524ba72bdc267b7/pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/IndexLoadingConfig.java#L305) look as follows 
```
        deserializer =
            stdDeserializer.withFallbackAlternative( IndexConfigDeserializer.fromMap(table -> fromIndexLoadingConfig));
```
instead of 
```
        deserializer =
            IndexConfigDeserializer.fromMap(table -> fromIndexLoadingConfig).withFallbackAlternative(stdDeserializer);
```

@gortiz Do you see any issues with making the above change?

**Comment by gortiz** [comment]
I've added a draft where I add a test that checks this specific case. If I understood the issue correctly, the test does not reproduce this issue. It seems that the onHeap is true.

It may be an error in my synthetic test or it may indicate that the issue is in the reader code, where the situation is not the same as expected in the synthetic case.

**Comment by vvivekiyer** [comment]
@gortiz  In the draft PR, the call to getting the column level config still uses the `stdDeserializer` (`indexType::getConfig`). As I stated in the RCA section,  issue happens only when we use IndexLoading::getDeserializer where the configs are fetched from` IndexType::fromIndexLoadingConfig()`. 

Here's a draft PR https://github.com/apache/pinot/pull/12260 that tests this case.

**Comment by vvivekiyer** [comment]
Confirmed with @gortiz through slack DM that the fix suggested in the "Issue Description" is good to go. 
I'll create a separate PR (instead of folding it inside https://github.com/apache/pinot/pull/12223)  for easier management - just in case the the change breaks any older configs that we've not considered. 
```
deserializer =
    stdDeserializer.withFallbackAlternative( IndexConfigDeserializer.fromMap(table -> fromIndexLoadingConfig));
```
@gortiz  please feel free to add any additional thoughts

cc: @siddharthteotia @walterddr @Jackie-Jiang

**Comment by walterddr** [comment]
browsing through these. I was wondering what's the reason behind having 2 places to configure indexes? 
`tableIndexingConfig` vs. `fieldConfigList.indexes`

granted that one is viewing from the indexing side (e.g. what're all the columns that have this index) and from the column side what are all the indexes that this column should encode).

but that seems error prompt to have "one take priority over another".

**Comment by vvivekiyer** [comment]
I can add my understanding based on the index-spi pep doc and @gortiz can add more thoughts. 

tableIndexConfig is the old/current way to define (most) indexing configs. Currently, indexing config is defined in multiple places (eg Forward index is specified in FieldConfig.encoding, but can also be disabled in FieldConfig.properties.forwardIndexDisabled). More such examples are in the document - https://docs.google.com/document/d/1L_GJ2tAgZVxI_Ssa59--Mso3mK_HeAS7s7G9261-sOE/edit. Going forward, the recommendation is to use `fieldConfigList.indexes ` as the single place to store all indexing information.

> but that seems error prompt to have "one take priority over another".

This is done for backward compatibility till all tableConfigs are migrated to the new model of using `fieldConfigList.indexes`

**Comment by walterddr** [comment]
> > but that seems error prompt to have "one take priority over another".
> 
> This is done for backward compatibility till all tableConfigs are migrated to the new model of using `fieldConfigList.indexes`

from what i can see `fieldConfigList.indexes` has been added for more than 3 years. 🤦‍♂️ so it is likely it is not going to go away soon.

**Comment by gortiz** [comment]
TL;DR:
1. `IndexingConfig.whatever` (where whatever is related to a single column index): Is the oldest implementation and should be avoided.
2. `FieldConfig.indexType` and `FieldConfig.indexTypes` is the first try to have a common place to configure indexes, but ended up being inconsistent and wasn't able to provide a nice way to configure indexes. It should *strongly* be avoided.
3. `FieldConfig.indexes` is the new standard way to configure single column indexes. It is centralized (all indexes of the same column are in the same place), expressive (indexes can be configured with complex objects) and easy to maintain from developer perspective (if the object can be understood by Jackson, no code needs to be added to serialize/deserialize).
4. In general, we fail if a config is specified in more than one place (ie `IndexingConfig.whatever` and `FieldConfig.indexes`). In order to do so, we use `withExclusiveAlternative`.
5. In some cases (like here in dictionaries) it is very difficult to detect in code whether the `IndexingConfig.whatever` was used or not, so in these cases we use `withFallbackAlternative`, but then the order matters. In this case , **we should give priority to the new `FieldConfig.indexes` than the other options**.

**Comment by walterddr** [comment]
^ this should be a good addition to the documentation piece IMO

**Comment by gortiz** [comment]
Do you mean public customer focus documentation or some dev documentation? The former is in a way already added. Pinot documentation has been rewritten to basically always use the new syntax (`FieldConfig.indexes`) and add expandable sections with the deprecated old syntax (`FieldConfig.indexeType`, `FieldConfig.indexeTypes` and `IndexingConfig.whatever`). At least changes are in master branch (see https://github.com/pinot-contrib/pinot-docs) but what I don't know is why that documentation is not already deployed (https://docs.pinot.apache.org/ is still showing the old docs).

**Comment by walterddr** [comment]
> Do you mean public customer focus documentation or some dev documentation? 

i meant docs.pinot.apache.org. i remember some time ago we moved to gitbook. CC @xiangfu0 who has more context
